man/FontXY.1
.\" SPDX-License-Identifier: MIT
.\"
.\" Copyright (c) 2020-2024 Donovan Watteau
.\"
.\" Permission is hereby granted, free of charge, to any person obtaining a copy
.\" of this software and associated documentation files (the "Software"), to deal
.\" in the Software without restriction, including without limitation the rights
.\" to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
.\" copies of the Software, and to permit persons to whom the Software is
.\" furnished to do so, subject to the following conditions:
.\"
.\" The above copyright notice and this permission notice shall be included in
.\" all copies or substantial portions of the Software.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
.\" IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
.\" FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
.\" AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
.\" LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
.\" OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
.\" THE SOFTWARE.
.Dd $Mdocdate: November 12 2022 $
.Dt FONTXY 1
.Os
.Sh NAME
.Nm FontXY
.Nd adjust glyph dimensions inside SCUMM fonts
.Sh SYNOPSIS
.Nm FontXY
.Cm i
.Ar CHARFILE
.Nm FontXY
.Cm o
.Ar CHARFILE
.Sh DESCRIPTION
The
.Nm
tool lets you adjust the relative size of each glyph (or character) inside a
SCUMM character file.
The mode of operation is chosen with one of the following options:
.Bl -tag -width Ds
.It Cm i
Import glyph dimensions from a
.Pa XY.txt
file into the specified
.Ar CHARFILE
block.
.It Cm o
Export glyph dimensions from the specified
.Ar CHARFILE
block into a
.Pa XY.txt
file.
.El
.Pp
.Ar CHARFILE
must be a
.Dq CHAR_*
SCUMM block, as extracted by
.Xr scummrp 1 .
.Pp
Glyph dimensions are represented as a series of
.Ql X;Y
coordinates, relative to zero, each glyph being separated by a new line.
Empty glyphs are represented as empty lines.
.Sh SEE ALSO
.Xr scummfont 1
.Sh HISTORY
The
.Nm
tool was written in 2006 by Thomas Combeleran for the ATP team,
and was open-sourced in 2020 under the MIT license.
.Sh CAVEATS
Only newline characters native to the current platform are accepted
(CRLF on Windows and LF on most other systems).
.Pp
.Pa XY.txt
pathname is hardcoded.


NEWS.md
## ScummTR 0.6.0 (202?-??-??)

### Bugfixes

- ScummTR: Fix a fatal "ERROR: Unknown function id 0x2E" error when trying to import/export text for the German releases of Indy3. This is because of an original issue in the official German releases of the game, where some Eszett characters were badly encoded.

### New features

- ScummTR: try to detect and warn when lines end with bogus `\013` sequences, which can cause some subtle issues with ScummVM or the original interpreters. These sequences would appear when the `-w` option was used for an initial export, but forgotten in a subsequent import.
- ScummTR: when doing an import, try to detect files with Windows CRLF end of lines, so that the `-w` option will be turned on, even if the user forgot to do so. (It's still recommended to explicitly add this option when working with such files, though.)
- ScummTR: add a new `-n` option, which prevents ScummTR from emitting any `;; ScummTR note:` comment. This may help some older derivative tools, such as ScummSpeaks or LanguageBundleBuilder, which don't expect them (since original ScummTR 0.4 never produced them).
- Various manual page and warning message improvements.

## ScummTR 0.5.1 (2022-02-27)

### Bugfixes

- ScummTR/ScummRP: Fix a fatal "Bad OIv2 offset" error with (at least) some versions of Maniac Mansion, Zak McKracken and Indy3. This regression was introduced in 0.5.0.

### Known bugs

- ScummTR/ScummRP: Mark all V2 versions of Maniac Mansion as unsupported, and not just the English V2 version. The French V2 version has the same corruption problem as the English version, for example.

## ScummTR 0.5.0 (2022-02-22)

### New features

- MS-DOS binaries have been added, on top of the macOS, Linux and Windows binaries.
- macOS binaries now have support for PowerPC and Apple silicon processors.
- Manual pages have been added (with .txt versions for systems where the original format is impractical).
- ScummTR: add a new `-r` option, which interprets the text of the game "as-is", in its original encoding. This can be useful for non-Latin languages such as Japanese.
- ScummTR: make it possible to import back a translation file created with the `-h` and/or `-I` options.
- ScummTR: add some encoding headers at the start of the translation file, so that most text editors will properly detect its encoding.
- ScummTR/ScummRP: make "XX.LFL should actually end at 0x1234" a warning which is always displayed, instead of being only printed in verbose mode.
- ScummFont: when importing a font, you now don't have to manually replace the original file with the `-new` file anymore.
- ScummFont: provide more details if an unsupported BMP file is given as an input.

### Bugfixes

- ScummTR/ScummRP: Fix a game corruption bug with at least LOOM-EGA (and possibly some earlier games too), such as clicking on the hole of the leftmost tree of the forest at the start of the game. In the official pre-built binaries, only `scummtr-0.4.2-win32.zip` was impacted by this corruption issue.
- ScummTR/ScummRP: fix a fatal "Duplicate offset in index" error with some versions of LOOM-EGA-EN.
- ScummTR/ScummRP: don't fail when the DISK09.LEC file is missing for MONKEY1-EGA, since it was only available through the Roland Update.
- ScummTR/ScummRP: don't fail and report missing files for the 4-disk floppy versions of MONKEY1, because only the 8-disk floppy versions need them.
- ScummTR/ScummRP: detect and reject the default version of MONKEY1-EGA that Limited Run Games sold in its Monkey Island Anthology; its DISK04.LEC file is known to be corrupted, but we have a guide describing how to recover it.
- ScummFont: reject more incompatible BMP files, instead of trying to use them and then silently corrupting the internal SCUMM fonts. Note that this makes ScummFont very strict; at the moment, it is recommended to use GIMP (with the "Do not write color space information" export option), or older versions of Microsoft Paint if you need to edit a BMP font.

### Incompatibilities with the previous version

- ScummTR: a translation file created by an earlier version of ScummTR will not necessarily be compatible with this new version. It is recommended to make a backup, import (`-i`) your current translation with the old ScummTR, and extract (`-o`) it again with the new ScummTR.
- ScummTR: the default value for the `-f` option is now the `scummtr.txt` file, instead of `./text`.
- ScummTR: the `-h` and `-I` options are now ignored, when the `-b` option is used.
- ScummRP: the default value for the `-d` option is now the `DUMP` directory, instead of the current (`.`) directory.
- ScummTR/ScummRP: unrecognized options now produce an immediate fatal error, instead of being silently ignored.

### Known bugs

- MANIAC-V2-EN is unsupported. This is because of an existing bug in the original game: any change made to the resources would amplify it and corrupt the game. Possible workaround: work from MANIAC-V1-EN or from a non-English version of Maniac Mansion V2.
- Some versions of MONKEY1-FLOPPY-VGA are currently unsupported. Possible workaround: work from a later version of MONKEY1-FLOPPY-VGA (it should display a higher number than "1.0" when doing Ctrl-V in-game).
- ScummTR: the `-b` option may not correctly work with all games. Possible workaround: use the `-r` option.

### Portability (for developers)

- Big-endian systems are now completely supported.
- Drop support for MSVC versions lower than 2015, because of their improper support for safe constructs such as `snprintf()`.
- Fix an incorrect usage of `std::sort()`, which could (at least) occur with g++ >= 4.5 and associated libstdc++.

## ScummTR 0.4.2 (2020-11-28)

### New features

- The original sources have been modified to be compatible with most modern systems.
- New binaries for Mac Intel systems.
- 64-bit binaries for Linux are now available too.

### Bugfixes

- None.

### Portability (for developers)


src/FontXY/FontXY.cpp
/*
 * SPDX-License-Identifier: MIT
 *
 * Copyright (c) 2006 Thomas Combeleran
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 * Note: the last "original" binary was built on: 2006-02-19 16:25:03.
 */

#include "common/types.hpp"
#include "common/file.hpp"
#include "common/toolbox.hpp"

#include <clocale>
#include <cstdlib>

#include <algorithm>
#include <fstream>
#include <iostream>
#include <list>
#include <string>
#include <vector>

int main(int argc, char **argv)
{
	std::setlocale(LC_CTYPE, "");

	if (argc != 3 || argv[1][0] == '\0' || argv[1][1] != '\0' || (argv[1][0] != 'i' && argv[1][0] != 'o'))
	{
		std::cout << "FontXY 0.6.0 (build " << SCUMMTR_BUILD_DATE << ") by Thomas Combeleran\n\n";
		std::cout << "Usage: FontXY {i|o} <CHAR file>\n\n";
		std::cout << "Examples:\tFontXY o CHAR_0002\n";
		std::cout << "         \tFontXY i CHAR_0003\n\n";
		printCommonDisclaimer();
		return 0;
	}

	bool bImport = argv[1][0] == 'i';
	char *pszChar = argv[2];
	char szTxt[] = "XY.txt"; // TODO: param

	if (bImport)
	{
		char szLine[1024];
		std::ifstream fTxt(szTxt, std::ifstream::in);
		File fChar;

		if (!fTxt.is_open())
		{
			std::cerr << "Error: cannot open " << szTxt << std::endl;
			return EXIT_FAILURE;
		}

		fChar.open(pszChar, std::ios::in | std::ios::out | std::ios::binary);
		if (!fChar.is_open())
		{
			std::cerr << "Error: cannot open " << pszChar << std::endl;
			return EXIT_FAILURE;
		}

		// Seek after the header
		fChar.seekg(8 + 0x17, std::ios::beg);

		int16 snNumChars; // <= 0x100
		fChar->getLE16(snNumChars);
		int nNumChars = snNumChars;
		if (nNumChars > 0x100 || nNumChars <= 0)
		{
			std::cerr << "Error: invalid number of chars: " << nNumChars << std::endl;
			return EXIT_FAILURE;
		}

		for (int i = 0; i < nNumChars; ++i)
		{
			uint32 uOffset;
			fChar.seekg(8 + 0x19 + i * 4, std::ios::beg);
			fChar->getLE32(uOffset);

			if (uOffset == 0)
			{
				fTxt.getline(szLine, sizeof szLine, '\n');
			}
			// Read the "x,y" line and update the CHAR file
			else
			{
				int nLeft, nTop;
				int8 byLeft, byTop;
				fTxt >> nLeft;
				fTxt.getline(szLine, sizeof szLine, ';');
				fTxt >> nTop;
				fTxt.getline(szLine, sizeof szLine, '\n');
				byLeft = (int8)nLeft;
				byTop = (int8)nTop;
				fChar.seekp(8 + 0x15 + uOffset + 2, std::ios::beg);
				fChar.write((char *)&byLeft, 1);
				fChar.write((char *)&byTop, 1);
			}
		}
		fTxt.close();
		fChar.close();
	}
	else
	{
		File fChar;
		std::ofstream fTxt(szTxt, std::ofstream::out | std::ofstream::trunc);

		if (!fTxt.is_open())
		{
			std::cerr << "Error: cannot open " << szTxt << std::endl;
			return EXIT_FAILURE;
		}

		fChar.open(pszChar, std::ios::in | std::ios::binary);
		if (!fChar.is_open())
		{
			std::cerr << "Error: cannot open " << pszChar << std::endl;
			return EXIT_FAILURE;
		}

		// Seek after the header
		fChar.seekg(8 + 0x17, std::ios::beg);

		int16 snNumChars; // <= 0x100
		fChar->getLE16(snNumChars);
		int nNumChars = snNumChars;
		if (nNumChars > 0x100 || nNumChars <= 0)
		{
			std::cerr << "Error: invalid number of chars: " << nNumChars << std::endl;
			return EXIT_FAILURE;
		}

		for (int i = 0; i < nNumChars; ++i)
		{
			uint32 uOffset;
			fChar.seekg(8 + 0x19 + i * 4, std::ios::beg);
			fChar->getLE32(uOffset);

			// Just add an empty line
			if (uOffset == 0)
			{
				fTxt << "\n";
			}
			// Read the values, and write them in the TXT file
			else
			{
				int8 byLeft;
				int8 byTop;
				fChar.seekg(8 + 0x15 + uOffset + 2, std::ios::beg);
				fChar.read((char *)&byLeft, 1);
				fChar.read((char *)&byTop, 1);
				fTxt << (int)byLeft << ";" << (int)byTop << "\n";
			}
		}
		fTxt.close();
		fChar.close();
	}
}
